### ISSUE TITLE
The equal-depth tie-break order is unwritten: QA can detect a change but not a wrong answer

### ISSUE THREAD

**Issue by klausi3D** [klausi3D/godotGS#792]
Split out of the #522 review. The QA lane now *detects a change* in the equal-depth tie-break order, but nobody has established which order is **correct**, so a reversal would be caught as "different" rather than "wrong".

## What the shader says

`modules/gaussian_splatting/shaders/tile_binning.glsl` packs the 32-bit key as:

```glsl
uint depth_key = (depth_quant << 8u) | (global_idx & 0xFFu);
```

So splats at bit-equal depth order by **ascending global index**. The stated purpose is determinism: *"Without the tie bits, splats in the same depth bucket sort by GPU atomic-emit completion order, which is non-deterministic across frames and produces dense-tile flicker on camera movement."*

## What the renderer does

`qa_sort_tie_breaker` places two splats at exactly the same position — splat 0 red, splat 1 green, `enable_tie_breaker=true`. Reading the ascending-index key naively, the higher index composites last and green should dominate the centre pixel.

Measured on an RTX 3090, bit-identical across three consecutive runs:

```
centre pixel   r=0.125  g=0.118
tie_break_winner = red
tie_break_margin = 0.00784314423799515
```

Red is ahead. Two things are worth noting before concluding anything:

1. **The margin is ~6%, not categorical.** Both splats are semi-transparent, so the centre is a *blend* of both contributions rather than a winner-takes-all pixel. The dominance direction is a real signal but a weak one.
2. **I could not establish the intended direction from the code.** Whether ascending key order means "drawn first" or "drawn last" in the composite depends on the rasterizer's traversal, which I did not trace.

So this is not filed as a bug. It is filed because the *expectation is unwritten*, and an unwritten expectation is how a reversal ships unnoticed.

## What #522 does in the meantime

Rather than assert a direction I could not ground, `qa_sort_tie_breaker` records the observed winner as `tie_break_winner` and the committed QA baseline pins it with the exact-match rule. A reversed tie-break flips `red` → `green` and the blocking lane fails. That closes the review finding — the scene previously asserted only frame-to-frame stability, which two consistently-misordered splats satisfy perfectly with SSIM 1.0 — without inventing a contract.

## What to do here

1. Trace the composite order and establish which splat *should* own the pixel for `key = (depth << 8) | index`. Document it next to the key packing.
2. Turn the recorded winner into an asserted one in `qa_sort_tie_breaker.gd` (there is a commented-out constant and the reasoning in place), so the scene fails on the *wrong* order rather than merely a *changed* one.
3. Consider whether two fully opaque splats would make a better fixture: the current blend gives a 6% margin, where opaque splats would give a categorical answer and a much stronger assertion.
4. While there: the same file notes a known limitation worth its own decision — at >5000 splats per active tile with only 256 tie values, birthday collisions give *"~56% chance of at least one within-bucket tie collision per (tile, bucket) pair, so some residual flicker remains"*. The suggested fix (reducing `GS_SORT_TILE_BITS` 16 → 13 to raise tie precision to 2048 values, ~10% collision) is described in the shader as a project-setting change that was out of scope for a shader-only fix.

**Comment by klausi3D** [comment]
## Human disposition packet — immutable master `873025f5783c765caec08a39a33909d1118b9d60`

No behavior, threshold, or baseline was changed while preparing this packet.

### Composite-order trace

The previously missing traversal trace is now complete on the live master SHA:

1. The shippable 64-bit configuration reserves tie bits and appends the low global-index bits to the depth key ([tile_renderer.cpp](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/renderer/tile_renderer.cpp#L2248-L2258), [tile_binning.glsl](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/shaders/tile_binning.glsl#L349-L356)). Equal depth therefore sorts by ascending global index.
2. The radix prefix walks bins from low to high and stable-scatter writes at that ascending prefix ([gpu_sorter.cpp](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/renderer/gpu_sorter.cpp#L1978-L1985), [scatter](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/renderer/gpu_sorter.cpp#L2090-L2103)).
3. The rasterizer consumes the sorted range from index 0 upward ([tile_rasterizer_compute.glsl](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/shaders/tile_rasterizer_compute.glsl#L229-L253)).
4. It uses front-to-back remaining-alpha accumulation: each earlier contributor consumes alpha available to every later one ([tile_raster_common.glsl](https://github.com/klausi3D/godotGS/blob/873025f5783c765caec08a39a33909d1118b9d60/modules/gaussian_splatting/shaders/includes/tile_raster_common.glsl#L446-L472)).

Therefore the current implementation means: **ascending global index is traversed first and receives first/front-to-back priority**. In the fixture, index 0 is red, so the observed red-leading blend is consistent with the code. The old “green composites last and should dominate” inference assumed back-to-front over-compositing and was incorrect for this rasterizer.

### Runtime evidence and remaining red

- The committed synthetic oracle records `tie_break_winner=red`, margin `0.00784314423799515`.
- The 2026-08-23 scheduled run on this exact master SHA kept the direction but measured margin `0.003922`; the unchanged guard requires `>= 0.005490`, so the lane correctly failed: [run 32617188157](https://github.com/klausi3D/godotGS/actions/runs/32617188157).
- The real-scan post-merge package uses the same 64-bit production path on `Cabin in the forest.ply` (~130k splats), RTX 3090, 1920×1080, scale 1.0/0.75 and FSR2 motion. It is under `C:\Projects\godotGS-remediation\phase2\gpu001-visual\runs\merged924_behavioral_4293__*`. Natural scan data does **not** provide a bit-equal-depth labeled pair, so it is valid downstream visual evidence but cannot choose the tie semantic. Pretending otherwise would be a vacuous oracle.

### Required written choices

Please reply with one value from each axis:

**Semantic**

- `SEMANTIC=KEEP_ASCENDING_INDEX_FIRST` — recommended from the current code/runtime trace; document and assert red as the categorical direction.
- `SEMANTIC=REVERSE_PRIORITY` — requires a separate R3 behavior change plus synthetic oracle and real-scan recapture; do not re-pin first.

**Current margin regression**

- `MARGIN=FIX_BEHAVIOR` — recommended while the unexplained ~50% signal collapse remains; keep the baseline and threshold unchanged.
- `MARGIN=SANCTION_REPIN` — explicit owner approval to prepare a separate reviewed baseline PR, with rationale and repeated exact-master captures. This reply is the required authorization; it does not itself change the baseline.

Until both choices are written, issue #792 and the master GPU lane remain intentionally blocked. No agent will infer the product semantic or re-pin from the present observation alone.

**Comment by klausi3D** [comment]
Closing: the unwritten expectation this issue was filed about is now written down and enforced.

## What landed

PR #956, squash-merged as `8be4a476b655b17f76252f95806e35abb62a95b5` on `master` (verified an ancestor of `origin/master`).

The equal-depth contract was traced on the **production 64-bit path** (`tile_binning.glsl:351-355`, not the 32-bit snippet quoted in the issue description): equal depths sort by ascending global index, the rasterizer traverses the sorted range from index zero upward, and front-to-back remaining-alpha accumulation gives the first contributor priority. Index 0 is red, so red must win.

`tests/examples/godot/test_project/scenes/qa/qa_sort_tie_breaker.gd` now asserts that answer instead of recording it:

- `EXPECTED_TIE_BREAK_WINNER := "red"`, checked in the scene's own pass/fail (`:157`), with the reasoning written out at `:13-24`.
- `MINIMUM_TIE_BREAK_MARGIN := 5.0 / 255.0`, so a directionally-correct-but-collapsed signal also fails.
- The fixture is made effectively opaque via `set_opacity_logits([8.0, 8.0])`. Root cause of the old weak margin: `set_splat_count()` allocates a zero-filled opacity-logit lane that decodes to `0.5` and takes precedence over `Color.a`, so the "opaque" splats were never opaque.
- The committed QA baseline and its exact-match rule are unchanged and remain an independent signal-collapse guard.

Against issue items: **item 2** (assert rather than detect) and **item 3** (opaque fixture, categorical answer) are done. **Item 1** is substantively done — the order was traced and documented — though the write-up lives in the QA scene rather than beside the key packing.

## Evidence

Required exact-head checks all passed on `00c22f103e`: `agentic-pr-gate`, `GPU QA Tests (Windows Self-Hosted)` 9/9, `Guards (Render Path + Static Safety)`, `GPU Harness + Visual Gate (Windows Self-Hosted)`, `GPU Evidence Requirement`, `Module Build + Runtime Harness (Windows Self-Hosted)`, `CodeFactor`.

Real Windows/Vulkan QA: 9 scenes, 152 metrics, 0 regressions, SSIM min/avg 1.0, `tie_break_winner=red`, `tie_break_margin=0.03137255` — 8x the `0.003922` that current-master produced against the `0.005490` floor before the fix. Two consecutive CI runs measured `6 / 255`.

Discriminating power was mutation-proven twice: flipping the expected winner `red` -> `green` failed the focused source guard, and deleting `asset.set_opacity_logits(...)` failed it as well.

Local: 83 focused tests, 132 agentic tests, `--guard-only`, and the task-contract check all passed. No renderer, shader, baseline, threshold, workflow, or release-gate file was changed.

## Carried forward

**Item 4** (the >5000-splats-per-tile birthday collision and the `GS_SORT_TILE_BITS` 16 -> 13 rebalance) is deliberately **not** closed here — this issue itself called it a separate decision. It is now #957, which also carries the leftover half of item 1 (a contract pointer next to the key packing, comment-only but R2 by path policy). #957 additionally records a reachability finding: `tie_bits` is only nonzero when `key_bits > 32` (`tile_renderer.cpp:2252-2258`), so the 32-bit tie branch the caveat describes may be unreachable at runtime and in the shader-compile CI, which would make the proposed rebalance work on dead code.
